Up(sun) and ready with Pandoc (#29)
upsun/devcenter: dev/content/posts/install-pandoc-on-upsun.md

a new standard has been proposed to the web ecosystem and will soon become essential for the web: [``llms.txt``](https://llmstxt.org/).
LLMs.txt was conceived by [Jeremy Howard](https://github.com/jph00), co-founder of Answer.AI, to address a fundamental challenge in AI-human interaction.

When AI assistants attempt to process standard web pages, they struggle with non-essential elements like navigation menus, scripts, and styling. 
These elements consume valuable context space without contributing to the actual content understanding.
LLMs.txt provides an elegant solution: it delivers precisely curated information in a format that AI systems can efficiently process and understand.

If you need to convert files from one markup format to another, [Pandoc](https://pandoc.org/) is your swiss-army knife.
Developed by [John MacFarlane](https://github.com/jgm), Pandoc is a [Haskell](https://haskell.org/) library for converting from one markup format to another 
and John provides in [this `pandoc` repo](https://github.com/jgm/pandoc) a command-line tool that uses this Pandoc library. 
Easy to install and ready to convert.

> In this How-Tos guide, we will see how to install this `pandoc` command line tool on all your Upsun projects.

Reviewer: ```suggestion
In this How-Tos guide, we will see how to install this `pandoc` command line tool on your Upsun project.
```

{{< callout >}}

**Assumptions:**

- You already have an Upsun account. If you don’t, [please register for a trial account](https://auth.upsun.com/register). You can sign up with an email address or an existing GitHub, Bitbucket, or Google account. If you choose one of these accounts, you can set a password for your Upsun account later.
- You have the [Upsun CLI](https://docs.upsun.com/administration/cli.html) installed locally.

{{< /callout >}}

{{< callout >}}

🚨 **Please note**: The processes/commands in the Up(sun) and running with Pandoc tutorial can be adapted for your Platform.sh project by replacing `upsun` CLI by `platform` CLI.

{{< /callout >}}

## Create an Upsun project

{{< callout >}}

🚨 **Please note**: for this tutorial, we use a [Symfony](https://www.symfony.com) application but feel free to use any other framework you like. 
If so, you can directly jump to next step: [Create a new environment](#create-a-new-environment) and replace ``symfony`` by ``upsun`` in all shell command-lines for the rest of this blogpost.

{{< /callout >}}

You first need a deployed project on Upsun. 

> If you don't, please follow all these steps from this tutorial [how to install a Symfony demo on Upsun](https://upsun.com/blog/upsun-and-running-with-symfony-demo/).

Reviewer: As you're not using anything related to Symfony, why not create a simple empty project using the `upsun` CLI instead? That would help having more people willing to try, right?

Author: Sorry, I'm too Symfony centric :D 
but as you did such amazing work to ease deploying Symfony project on Upsun, I though it could be great to promote my previous blogpost.
I will try to update it to remove Symfony specifics.

Or execute these few command lines:
```shell
symfony new my-symfony-app --demo --upsun
cd my-symfony-app
symfony project:create
symfony deploy
```

## Install Pandoc

{{% steps %}}

### Create a new environment

> As we never (ever) test new features directly on production, we will create a new [environment](https://docs.upsun.com/environments.html) to test Pandoc installation.

Reviewer: As you're probably going to create a dev project anyway in the step above, I would not create a new env as this is not really useful for this tutorial (it just adds one more step).

Author: yes, but the goal of this step is too show that creating a new env to experiment is possible (and quick) with Upsun.
But you're highlighting that i miss the "merge to prod" step at the end.

Reviewer: To me, a blog post should be focus on one thing, no need to reiterate all the good stuff we have :)

To create a new environment on your project, use the following command:

```shell
upsun branch pandoc-install 
```

This command creates a new preview environment and automatically switches your local Git branch to ``pandoc-install``.

### Use a shell script to install Pandoc

[John MacFarlane](https://github.com/jgm) provides in his [Pandoc repo](https://github.com/jgm/pandoc) a quick and easy way to install Pandoc.

> We design a shell script for you ([source](https://github.com/upsun/snippets/blob/main/src/install-pandoc.sh)) that can be used to install Pandoc (ony once on every new releases) 

Reviewer: Is it possible to use the composable image instead?

Author: Composable Image (CI)? 
i didn't try as this feature is on the side/paused for now (or I am missing something?) and someone told me that we don't want to "loose" more time on it...
Do you want me to spend some time to experiment using CI?
Other than that, yes, we can [use CI](https://search.nixos.org/packages?channel=24.11&show=pandoc_3_5&from=0&size=50&sort=relevance&type=packages&query=pandoc) to install it

during the ``applications.app.hooks.build`` step:

```yaml
applications:
  app:        
    #...
    hooks:
      build: |
        set -x -e
        #...
        curl -fsS https://raw.githubusercontent.com/upsun/snippets/refs/heads/main/src/install-pandoc.sh | bash
      deploy: |
        set -x -e
        #...
        bin/pandoc -v
```

This ``install-pandoc.sh`` script installs the ``pandoc`` binary, from [Pandoc repo](https://github.com/jgm/pandoc), in the `/app/bin` folder of your application container.

### Deploy your updates.

```shell
git add .upsun/config.yaml && git commit -m "install Pandoc during build"
symfony deploy #or upsun push
```

### Use it
You can now use ``pandoc`` in your project:
```shell
applications:
  app:        
    #...
    hooks:
      build: |
        set -x -e
        #...
        curl -fsS https://raw.githubusercontent.com/upsun/snippets/refs/heads/main/src/install-pandoc.sh | bash
      deploy: |
        set -x -e
        #...

>         bin/pandoc $(find ./public -iname "*.html" -type f | sort -d) -f html -s -o "var/cache/llms.txt" -t markdown

Author: i change the entire blogpost to "simulate" a simple HTML applications and better explain what I am doing when using pandoc to create the llms.txt.

```

And deploy your changes: 

```shell
git add .upsun/config.yaml && git commit -m "use Pandoc to concat all public .html files to var/cache/llms.txt"
symfony deploy #or upsun push
```

{{< callout >}}

🚨 **Please note**: If you use `pandoc` binary during ``hooks.deploy`` step, please be sure your destination folder (here ``/app/var/cache``) is writable. 
Please see [Mounts doc page](https://docs.upsun.com/create-apps/app-reference/single-runtime-image.html#mounts) for more info.

{{< /callout >}}

{{% /steps %}}

## Conclusion
Et voilà, you can now use ``pandoc`` to convert all existing public HTML pages in a single Markdown file `llms.txt` and start feeding AI Assistants with it.

> I keep you posted, some more blogposts on that subject are coming.

Reviewer: Let's remove that line

If you need to convert files from one markup format to another, [Pandoc](https://pandoc.org/) is your swiss-army knife.
Developed by [John MacFarlane](https://github.com/jgm), Pandoc is a [Haskell](https://haskell.org/) library for converting from one markup format to another 
and John provides in [this `pandoc` repo](https://github.com/jgm/pandoc) a command-line tool that uses this Pandoc library. 
Easy to install and ready to convert.

In this How-Tos guide, we will see how to install this `pandoc` command line tool on your Upsun project.

{{< callout >}}

**Assumptions:**

- You already have an Upsun account. If you don’t, [please register for a trial account](https://auth.upsun.com/register). You can sign up with an email address or an existing GitHub, Bitbucket, or Google account. If you choose one of these accounts, you can set a password for your Upsun account later.
- You have the [Upsun CLI](https://docs.upsun.com/administration/cli.html) installed locally.
- You have the [Git CLI](https://github.com/git-guides/install-git) installed locally.

{{< /callout >}}

{{< callout >}}

> 🚨 **Please note**: The processes/commands in this Up(sun) and running with Pandoc tutorial can be adapted for your Platform.sh project by replacing `upsun` CLI with `platform` CLI.

Reviewer: I would recommend removing this note as we want people to test and use Upsun.

{{< /callout >}}

> For this tutorial, we will start from a basic HTML application, but feel free to use any other framework you like.

Reviewer: ```suggestion
For this tutorial, we will start with a basic HTML application.
```

The main goal of this tutorial is to showcase how to install ``pandoc`` on your project and quickly generate a ``llms.txt`` file from your HTML pages.

> Your application needs to generate static HTML files in the ``public`` directory, like [Hugo](https://gohugo.io/).  

Reviewer: Kind of duplicate of the following paragraph

```suggestion
```

## Prepare your local HTML project

> In order to quickly showcase the strength of Pandoc, we will simulate a simple HTML application, that you could obtain using a stack like [Hugo](https://hugo.io/).

Reviewer: ```suggestion
In order to quickly showcase the strength of Pandoc, we will simulate a simple HTML application, that you could obtained using a static website generator like [Hugo](https://hugo.io/).
```

The proposed structure will be: 
```txt
├── .upsun
│   └── config.yaml
└── public
    ├── learn
    │   ├── api.html
    │   └── applications.html
    └── index.html     
```

> To do so, in your Terminal, execute the following commands:

Reviewer: Let's replace this with a ZIP file that already contains these files.

Author: done

What language is your project using? We support the following: [JavaScript/Node.js]

Tell us your project's application name: [app]

                       (\_/)
We’re almost done...  =(^.^)=

Last but not least, unless you’re creating a static website, your project uses services. Let’s define them:

Select all the services you are using: []

You have not selected any service, would you like to proceed anyway? [Yes]

┌───────────────────────────────────────────────────┐
│   CONGRATULATIONS!                                │
│                                                   │
│   We have created the following files for your:   │
│     - .environment                                │
│     - .upsun/config.yaml                          │
│                                                   │
│   We’re jumping for joy! ⍢                        │
└───────────────────────────────────────────────────┘
         │ /
         │/
         │
  (\ /)
  ( . .)
  o (_(“)(“)
```
Please select 
- ``Javascript/Node.js``
- application name: ``app``
- no service selected

Your HTML application is almost ready to be deployed on Upsun, one more step to go. 

> You just need to update this config line into your newly created ``.upsun/config.yaml`` for the router to point on your ``public`` folder:

Reviewer: ```suggestion
Update this config line into the newly created ``.upsun/config.yaml`` file for the router to point to your ``public`` folder:
```

```yaml {filename=".upsun/config.yaml"}
applications:
  app:
    web:
      locations:
        "/":
          root: "public"
          index: ["index.html"]
          passthru: true
```

and then commit your updates: 
```shell {filename="Terminal"}
git add .upsun/config.yaml && git commit -m "change locations.root to the public folder"
```

## Create an Upsun project

> You then need a create an Upsun project by executing these command line (and follow the prompts):

Reviewer: ```suggestion
You then need to create an Upsun project by executing these commands and follow the prompts:
```

```shell {filename="Terminal"}
upsun project:create
upsun deploy
```

## Install Pandoc

{{% steps %}}

### Use a shell script to install Pandoc

[John MacFarlane](https://github.com/jgm) provides in his [Pandoc repo](https://github.com/jgm/pandoc) a quick and easy way to install Pandoc.

> We design a shell script for you ([source](https://github.com/upsun/snippets/blob/main/src/install-pandoc.sh)) that can be used to install Pandoc (only once on every new release). 

Reviewer: ```suggestion
We've prepared a shell script for you ([source](https://github.com/upsun/snippets/blob/main/src/install-pandoc.sh)) that can be used to install the latest version of Pandoc.
```

> Update your ``.upsun/config.yaml`` file and add this curl call in your ``applications.app.hooks.build`` step:

Reviewer: I still think that using the composable image would be better.

Author: Do you want me to split this part in 2 different part ? 

```
### Install Pandoc
...
#### Using a shell script
...

#### Using Composable image
...
```
This will change a bit the rest of the article (if CI, we are not using `bin/pandoc` but `pandoc`)

```yaml {filename=".upsun/config.yaml"}
applications:
  app:        
    #...
    hooks:
      build: |
        set -x -e
        #...
        curl -fsS https://raw.githubusercontent.com/upsun/snippets/refs/heads/main/src/install-github-asset.sh | bash /dev/stdin "jgm/pandoc" 
        bin/pandoc -v
```

> This [``install-pandoc.sh``](https://github.com/upsun/snippets/blob/main/src/install-pandoc.sh) script installs the ``pandoc`` binary, from [Pandoc repo](https://github.com/jgm/pandoc), in the `/app/bin` folder of your application container.

Reviewer: ```suggestion
The [``install-pandoc.sh``](https://github.com/upsun/snippets/blob/main/src/install-pandoc.sh) script installs the ``pandoc`` binary from [Pandoc repo](https://github.com/jgm/pandoc) in the `/app/bin` folder of your application container.
```

> ### Deploy your updates.

Reviewer: ```suggestion
### Deploy your updates
```

```shell  {filename="Terminal"}
git add .upsun/config.yaml .environment && git commit -m "install Pandoc"
upsun push
```

### Create a `llms.txt` file

> You can now use ``pandoc`` in your project to generate a ``public/llms.txt`` file that will concat all of your HTML pages in Markdown format.

Reviewer: ```suggestion
You can now use ``pandoc`` in your project to generate a ``public/llms.txt`` file that will concatenate all the HTML pages in Markdown.
```

> Update your ``.upsun/config.yaml`` and add the following lines:

Reviewer: ```suggestion
Update your ``.upsun/config.yaml`` by adding the following lines:
```

```yaml {filename=".upsun/config.yaml"}
applications:
  app:        
    #...
    hooks:
      build: |
        set -x -e
        #...
        curl -fsS https://raw.githubusercontent.com/upsun/snippets/refs/heads/main/src/install-github-asset.sh | bash /dev/stdin "jgm/pandoc"
        bin/pandoc $(find ./public -iname "*.html" -type f | sort -d) -f html -s -o "./public/llms.txt" -t markdown
```

> This new command line ``bin/pandoc ...`` will concat all existing ``.html`` files, located in the `public` folder in a single `./public/llms.txt` file and convert them in a Markdown format.

Reviewer: ```suggestion
This ``bin/pandoc ...`` call will concatenate all existing ``.html`` files located in the `public` folder in a single `./public/llms.txt` file and convert them to Markdown.
```

{{< callout >}}

> 🚨 **Please note**: This ``install-github-asset.sh`` script accepts a second parameter defining the exact `<org>/<repo>` release version you want to install the asset from.
>    As we didn't provide any in our case, it will automatically install it from the latest release founded.

Reviewer: Let's keep things simple

```suggestion
```

{{< /callout >}}

> And then, you just need to deploy your updates: 

Reviewer: We should avoid expressions like `you just need to`. Sometimes these things are not that simple.
```suggestion
And then, deploy your updates: 
```

```shell {filename="Terminal"}

> git add .upsun/config.yaml && git commit -m "use Pandoc to concat all public .html files to a public/llms.txt file"

Reviewer: Let's use best practices for the Git commit messages:

```suggestion
git add .upsun/config.yaml && git commit -m "Use Pandoc to generate a public/llms.txt file"
```

upsun push
```

> You can then access this file by adding ``/llms.txt`` to your environment url:

Reviewer: ```suggestion
Test it works by accessing the file by adding ``/llms.txt`` to your environment URL:
```

```shell {filename="Terminal"}
upsun env:url --primary
```

> {{< callout >}}
> 
> 🚨 **Please note**: If you use `pandoc` tool to generate Markdown files during the ``hooks.deploy`` step, please make sure your destination folder is writable. 
> Please see [Mounts doc page](https://docs.upsun.com/create-apps/app-reference/single-runtime-image.html#mounts) for more info.
> 
> {{< /callout >}}

Reviewer: ```suggestion
```

{{% /steps %}}

## Conclusion

> Et voilà, we saw how easy it is to use ``pandoc`` tool to convert all your existing HTML pages in a single Markdown file `public/llms.txt`. 

Reviewer: ```suggestion
Et voilà, we saw how to use ``pandoc`` to convert all and existing set of HTML pages in a single Markdown `public/llms.txt` file.
```

Next step would be to train AI Assistants with this new file ``llms.txt``?

> Discover how easy it is deploying a personal [Chainlit](https://github.com/Chainlit/chainlit) AI assistant on Upsun by following this great blogpost: [Experiment with Chainlit AI interface with RAG on Upsun](https://devcenter.upsun.com/posts/deploying-chainlit-with-rag/) 

Reviewer: ```suggestion
Discover how to deploy a personal [Chainlit](https://github.com/Chainlit/chainlit) AI assistant on Upsun by reading this great blogpost: [Experiment with Chainlit AI interface with RAG on Upsun](https://devcenter.upsun.com/posts/deploying-chainlit-with-rag/) 
```

> Remain up-to-date on the latest from us over on our social media and community channels: [DEV.to](https://dev.to/upsun), [Reddit](https://www.reddit.com/r/upsun/), and [Discord](https://discord.gg/platformsh). 

Reviewer: Should be reworded :) @Kemi-Elizabeth Can you help here please?

      build: |
        set -x -e
        pandoc -v
```

And then, deploy your updates:

```shell  {filename="Terminal"}
git add .upsun/config.yaml .environment && git commit -m "install Pandoc"
upsun push
```

{{% /steps %}}

## Test it
You can now use ``pandoc`` in your project to generate a ``public/llms.txt`` file that will concatenate all the HTML pages in Markdown.
Update your ``.upsun/config.yaml`` by adding the following lines:
```yaml {filename=".upsun/config.yaml"}
applications:
  app:        
    #...
    hooks:
      build: |
        set -x -e
        #...
        pandoc $(find ./public -iname "*.html" -type f | sort -d) -f html -s -o "./public/llms.txt" -t markdown
```

This ``pandoc $(find...`` command concatenates all existing ``.html`` files located in the `public` folder in a single `./public/llms.txt` file and convert them to Markdown.

And then, deploy your updates: 

```shell {filename="Terminal"}
git add .upsun/config.yaml && git commit -m "Use Pandoc to generate a public/llms.txt file"
upsun push
```

Test it works by accessing the file by adding ``/llms.txt`` to your environment URL:
```shell {filename="Terminal"}
upsun env:url --primary
```

## Conclusion

> Et voilà, we saw how to use ``pandoc`` to convert all and existing set of HTML pages in a single Markdown `public/llms.txt` file.

Reviewer 2: ```suggestion
Et voilà, we saw how to use ``pandoc`` to convert all existing HTML pages into a single Markdown `public/llms.txt` file. Now, perhaps the next step would be to train an AI Assistant with the file ``llms.txt``...
```
